### ISSUE TITLE
Bad PASV/EPSV response: 229 for DIRLISTONLY

### ISSUE THREAD

**Issue by ribtoks** [curl/curl#10598]
### I did this

I'm running one of the basic libcurl examples to get listing of the server root directory, something along the lines of below (redacted a bit).

> sorry, not 100% "complete" example, because it's cut from the larger codebase and this issue is created from my app's user complaint, which I'm trying to debug to help that user

```
curlHandle = curl_easy_init();

curl_easy_setopt(curlHandle, CURLOPT_URL, ....);
curl_easy_setopt(curlHandle, CURLOPT_USERNAME, ....);
curl_easy_setopt(curlHandle, CURLOPT_PASSWORD, ....);

curl_easy_setopt(curlHandle, CURLOPT_WRITEFUNCTION, discardfunc);
curl_easy_setopt(curlHandle, CURLOPT_READFUNCTION, readfunc);
curl_easy_setopt(curlHandle, CURLOPT_DEBUGFUNCTION, my_trace);
curl_easy_setopt(curlHandle, CURLOPT_USERAGENT, ....);
// was changed in curl 7.74.0 to default "1"
curl_easy_setopt(curlHandle, CURLOPT_FTP_SKIP_PASV_IP, 0L);

curl_easy_setopt(curlHandle, CURLOPT_UPLOAD, 0L);
curl_easy_setopt(curlHandle, CURLOPT_DIRLISTONLY, 1L);
curl_easy_setopt(curlHandle, CURLOPT_HEADERFUNCTION, throw_away);
curl_easy_setopt(curlHandle, CURLOPT_FTP_USE_EPSV, 0L);
curl_easy_setopt(curlHandle, CURLOPT_VERBOSE, 1L);

r = curl_easy_perform(curlHandle);
```
Here's the trace of what happened:

```
13:38:20.252 T#554403 Connectivity::isContextValid - About to check credentials for "ftp://upload.alamy.com"
13:38:20.359 T#554403 Connectivity::my_trace - "== Info: Trying 18.200.74.74:21..."
13:38:20.508 T#554403 Connectivity::my_trace - "== Info: Connected to upload.alamy.com (18.200.74.74) port 21 (#0)"
13:38:20.747 T#554403 Connectivity::dump - "<= Recv header, 0000000007 bytes (0x00000007)\n220-*.."
13:38:20.747 T#554403 Connectivity::dump - "<= Recv header, 0000000051 bytes (0x00000033)\n220 This is a private system - No anonymous login.."
13:38:20.747 T#554403 Connectivity::dump - "=> Send header, 0000000029 bytes (0x0000001d)\nUSER <redacted>.."
13:38:20.908 T#554403 Connectivity::dump - "<= Recv header, 0000000055 bytes (0x00000037)\n331 User <redacted> OK. Password required.."
13:38:20.908 T#554403 Connectivity::dump - "Password token hidden here"
13:38:21.532 T#554403 Connectivity::dump - "<= Recv header, 0000000043 bytes (0x0000002b)\n230 OK. Current restricted directory is /.."
13:38:21.532 T#554403 Connectivity::dump - "=> Send header, 0000000005 bytes (0x00000005)\nPWD.."
13:38:21.747 T#554403 Connectivity::dump - "<= Recv header, 0000000034 bytes (0x00000022)\n257 \"/\" is your current location.."
13:38:21.747 T#554403 Connectivity::my_trace - "== Info: Entry path is '/'"
13:38:21.747 T#554403 Connectivity::my_trace - "== Info: Request has same path as previous transfer"
13:38:21.747 T#554403 Connectivity::dump - "=> Send header, 0000000006 bytes (0x00000006)\nPASV.."
13:38:21.747 T#554403 Connectivity::my_trace - "== Info: Connect data stream passively"
13:38:21.747 T#554403 Connectivity::my_trace - "== Info: ftp_perform ends with SECONDARY: 0"
13:38:21.907 T#554403 Connectivity::dump - "<= Recv header, 0000000047 bytes (0x0000002f)\n229 Extended Passive Mode Entered (|||37937|).."
13:38:21.907 T#554403 Connectivity::my_trace - "== Info: Bad PASV/EPSV response: 229"
13:38:21.907 T#554403 Connectivity::my_trace - "== Info: Remembering we are in dir \"\""
13:38:21.907 T#554403 Connectivity::my_trace - "== Info: Connection #0 to host upload.alamy.com left intact" 
```

### I expected the following

After the login succeeded (from the logs), I expected directory listing to succeed too.

### curl/libcurl version

7.86.0

### operating system

macOS 10.16 (Big Sur)

**Comment by bagder** [comment]
The FTP server seems confused. We can see curl sending a `PASV` but getting a 229 response back instead of the expected 227.

The 229 response is what a server is expected to return to `EPSV` command. You have (for some reason), disabled the use of the EPSV command. The 229 response looks like a perfectly valid 229 response, but that is not what a client expects to get back from a PASV.

**Comment by ribtoks** [comment]
Thank you for fast response.

EPSV was disabled for directory listing because it proved not to work in past (code was the same, but EPSV easy opt was `1L`). "Not to work" in this context means timeout. I don't know if it makes any sense, but this is what I observed (running this software with libcurl since 2014).

This error started to show up for 3 different companies that run FTP servers. As I can assume, might be they have 1 vendor that has updated their FTP server software.

Is there something I do incorrectly? Do you think there's a reason server would expect me to turn on EPSV, given current code? ("usual" passive mode is ON)

**Comment by bagder** [comment]
> might be they have 1 vendor that has updated their FTP server software.

Do you have any clue as to what server software/version they use? I'm thinking that if this behavior becomes common then we really need to adapt to it - as opposed to just being a single weirdo.

> Is there something I do incorrectly?

No. This is the server sending back a totally unexpected response. The only we way to make libcurl handle this situation is to make it more liberal in what response it should accept for `PASV`., It's not even difficult. If you're up for it, I could provide a test patch for you to see if we can make curl play nicely with this setup?

**Comment by bagder** [comment]
~~~diff
diff --git a/lib/ftp.c b/lib/ftp.c
index 7766f76c7..a5d388201 100644
--- a/lib/ftp.c
+++ b/lib/ftp.c
@@ -1807,12 +1807,11 @@ static CURLcode ftp_state_pasv_resp(struct Curl_easy *data,
   char *str = &data->state.buffer[4];  /* start on the first letter */
 
   /* if we come here again, make sure the former name is cleared */
   Curl_safefree(ftpc->newhost);
 
-  if((ftpc->count1 == 0) &&
-     (ftpcode == 229)) {
+  if(ftpcode == 229) {
     /* positive EPSV response */
     char *ptr = strchr(str, '(');
     if(ptr) {
       unsigned int num;
       char separator[4];
@@ -1851,12 +1850,11 @@ static CURLcode ftp_state_pasv_resp(struct Curl_easy *data,
     if(!ptr) {
       failf(data, "Weirdly formatted EPSV reply");
       return CURLE_FTP_WEIRD_PASV_REPLY;
     }
   }
-  else if((ftpc->count1 == 1) &&
-          (ftpcode == 227)) {
+  else if(ftpcode == 227) {
     /* positive PASV response */
     unsigned int ip[4] = {0, 0, 0, 0};
     unsigned int port[2] = {0, 0};
 
     /*

**Comment by ribtoks** [comment]
I don't know what exactly they run. Lately many of them switched to SFTP wrapper over S3, which is run by AWS ("managed" FTP), but I don't know exactly about 3 in question (Dreamstime, Alami & 123RF).

I'm up for the patch (Thank you!) and I will include it in the next hot fix release of my software. It will take a bit of time though to see the broad results, due to the time it takes to update users.

I know libcurl has like a million of configuration options already, but might be that this can become the million and 1st? (being disabled by default, of course)

**Comment by ribtoks** [comment]
An update: 1 of them runs Pure-FTPdd

**Comment by bagder** [comment]
I checked their source code and it responds with a 227 for `PASV`. Which does not match the behavior you saw.

https://github.com/jedisct1/pure-ftpd/blob/201bf0c31c33c0f1750642ba725e404f707ae41a/src/ftp_parser.c#L476

**Comment by ribtoks** [comment]
```

13:39:25.067 T#554408 Connectivity::my_trace - "== Info: Connected to ftp.depositphotos.com (135.84.53.98) port 21 (#0)"
13:39:25.329 T#554408 Connectivity::dump - "<= Recv header, 0000000063 bytes (0x0000003f)\n220---------- Welcome to Pure-FTPd [privsep] [TLS] ----------.."
13:39:25.330 T#554408 Connectivity::dump - "<= Recv header, 0000000046 bytes (0x0000002e)\n220-You are user number 75 of 10000 allowed..."
13:39:25.330 T#554408 Connectivity::dump - "<= Recv header, 0000000047 bytes (0x0000002f)\n220-Local time is now 19:39. Server port: 21..."
13:39:25.330 T#554408 Connectivity::dump - "<= Recv header, 0000000051 bytes (0x00000033)\n220-This is a private system - No anonymous login.."
13:39:25.330 T#554408 Connectivity::dump - "<= Recv header, 0000000055 bytes (0x00000037)\n220-IPv6 connections are also welcome on this server..."
13:39:25.330 T#554408 Connectivity::dump - "<= Recv header, 0000000061 bytes (0x0000003d)\n220 You will be disconnected after 2 minutes of inactivity..."
13:39:25.330 T#554408 Connectivity::dump - "=> Send header, 0000000018 bytes (0x00000012)\nUSER redacted.."
13:39:25.637 T#554408 Connectivity::dump - "<= Recv header, 0000000044 bytes (0x0000002c)\n331 User redacted OK. Password required.."
13:39:25.637 T#554408 Connectivity::dump - "Password token hidden here"
13:39:25.988 T#554408 Connectivity::dump - "<= Recv header, 0000000040 bytes (0x00000028)\n230-This server supports FXP transfers.."
13:39:25.988 T#554408 Connectivity::dump - "<= Recv header, 0000000043 bytes (0x0000002b)\n230 OK. Current restricted directory is /.."
13:39:25.988 T#554408 Connectivity::dump - "=> Send header, 0000000005 bytes (0x00000005)\nPWD.."
13:39:26.187 T#554408 Connectivity::dump - "<= Recv header, 0000000034 bytes (0x00000022)\n257 \"/\" is your current location.."
13:39:26.188 T#554408 Connectivity::my_trace - "== Info: Entry path is '/'"
13:39:26.188 T#554408 Connectivity::my_trace - "== Info: Request has same path as previous transfer"
13:39:26.188 T#554408 Connectivity::dump - "=> Send header, 0000000006 bytes (0x00000006)\nPASV.."
13:39:26.188 T#554408 Connectivity::my_trace - "== Info: Connect data stream passively"
13:39:26.188 T#554408 Connectivity::my_trace - "== Info: ftp_perform ends with SECONDARY: 0"
13:39:26.395 T#554408 Connectivity::dump - "<= Recv header, 0000000047 bytes (0x0000002f)\n229 Extended Passive Mode Entered (|||35919|).."
13:39:26.395 T#554408 Connectivity::my_trace - "== Info: Bad PASV/EPSV response: 229"
13:39:26.395 T#554408 Connectivity::my_trace - "== Info: Remembering we are in dir \"\""
13:39:26.395 T#554408 Connectivity::my_trace - "== Info: Connection #0 to host ftp.depositphotos.com left intact"
13:39:26.395 T#554408 Connectivity::isContextValid - Credentials checking error: FTP: unknown PASV reply
```

**Comment by bagder** [comment]
This is still without the patch, right?

**Comment by ribtoks** [comment]
Without the patch so far. Just more user logs that I found.

**Comment by ribtoks** [comment]
Just to set expectations: I will try the patch first thing tomorrow morning, as I'm traveling atm. (Checking logs & commenting from GitHub Mobile app)

**Comment by ribtoks** [comment]
I did testing and it does work for me _without_ the patch (with my personal account at Depositphotos). I contacted my user and it does not work for them. My app (and libcurl) is built on Intel and run on M1 mac via Rosetta. I wonder if it can change anything...

**Comment by bagder** [comment]
> I did testing and it does work for me without the patch

Is this unprotected FTP? Ie no TLS? Could it be that this user has some weird firewall/program that interferes?

> I contacted my user and it does not work for them. 

Still unpatched you mean? Isn't that what this bug report is about?

**Comment by ribtoks** [comment]
It is an unprotected FTP (not SFTP and not FTPS). I actually asked the user if they have some firewall etc., but the user is not so advanced computer user. The user reported that FileZilla worked for them at the same time as my software with libcurl has not.

I will try to get my hands on M1 mac to test next week. Wonder if Rosetta emulation layer could change anything that much.

> Still unpatched you mean? Isn't that what this bug report is about?

After I verified that it works with my account, I thought it might be a transient issue on the (remote) company side for whatever reason, but it did not work for the user at the same time it worked for me.

**Comment by bagder** [comment]
I cannot fathom that this is at all related to CPU/architecture and m1 or x86 will not matter. This is the server responding with an unexpected reply. "Something" outside of curl is making this happen.

**Comment by ribtoks** [comment]
Finally I got more info from the client. They had this error with unencrypted FTP using T-Mobile's 5G connection. After they switched to a different provider using "usual" cables, the issue went away. Also they reported that another company that uses FTPS always worked for them. From everything it appears that provider was tinkering with the traffic, which is not nice.

Also I tried the M1 mac, but it worked without problems for my accounts.

Bottom line, I don't think there's anything curl should do differently. Waiting for your input, but I guess we can close the issue.
